src/plots/linePlot.ts

```diff
@@ -425,6 +449,71 @@ export module Plots {
       return filteredDataIndices;
     }
 
+    private _filterDownsampling(dataset: Dataset, indices: number[]) {
+      let data = dataset.data();
+      let filteredIndices: number[] = [];
+
+      let scaledXAccessor = Plot._scaledAccessor(this.x());
+      let scaledYAccessor = Plot._scaledAccessor(this.y());
+
+      if (indices.length === 0) {
+        return filteredIndices;
+      }
+      filteredIndices.push(indices[0]);
+
+      let indexBelongsToCurrentBucket = (i: number, currentSlope: number) => {
+        let p1x = scaledXAccessor(data[indices[i]], indices[i], dataset);
+        let p1y = scaledYAccessor(data[indices[i]], indices[i], dataset);
+        let p2x = scaledXAccessor(data[indices[i + 1]], indices[i + 1], dataset);
+        let p2y = scaledYAccessor(data[indices[i + 1]], indices[i + 1], dataset);
+        if (currentSlope === Infinity) {
+          return Math.floor(p1x) === Math.floor(p2x);
+        } else {
+          let expectedP2y = p1y + (p2x - p1x) * currentSlope;
+          return Math.floor(p2y) === Math.floor(expectedP2y);
+        }
+      };
+
+      for (let i = 0; i < indices.length - 1; ) {
+        let indexFirst = indices[i];
+        let p1x = scaledXAccessor(data[indices[i]], indices[i], dataset);
+        let p1y = scaledYAccessor(data[indices[i]], indices[i], dataset);
+        let p2x = scaledXAccessor(data[indices[i + 1]], indices[i + 1], dataset);
+        let p2y = scaledYAccessor(data[indices[i + 1]], indices[i + 1], dataset);
+        let currentSlope = (Math.floor(p1x) === Math.floor(p2x)) ? Infinity : (p2y - p1y) / (p2x - p1x);
+        let indexMin = indices[i];
+        let minScaledValue = (currentSlope === Infinity) ? p1y : p1x;
+        let indexMax = indexMin;
+        let maxScaledValue = minScaledValue;
+
+        while (i < indices.length - 1 && indexBelongsToCurrentBucket(i, currentSlope)) {
+          let currPoint = currentSlope === Infinity ? scaledYAccessor(data[indices[i + 1]], indices[i + 1], dataset) :
+            scaledXAccessor(data[indices[i + 1]], indices[i + 1], dataset);
+          if (currPoint > maxScaledValue) {
+            maxScaledValue = currPoint;
+            indexMax = indices[i + 1];
+          }
+          if (currPoint < minScaledValue) {
+            minScaledValue = currPoint;
+            indexMin = indices[i + 1];
+          }
+          i++;
+        }
+
+        let indexLast = indices[i];
+
+        if (indexMin !== indexFirst) {
+          filteredIndices.push(indexMin);
+        }
+        if (indexMax !== indexMin && indexMax !== indexFirst) {
+          filteredIndices.push(indexMax);
+        }
```

Comment:
Because the lines can go upward and downward. 
For example, we have a line with slope 1.
(1,1)(0,0),(2,2) are on the same line. But the min point is (0,0), not the first point.